File: Agents/Intelligent.js
export const generateText = async () => {
	const url = "https://us-south.ml.cloud.ibm.com/ml/v1/text/generation?version=2023-05-29";
	const headers = {
		"Accept": "application/json",
		"Content-Type": "application/json",
		"Authorization": "Bearer eyJraWQiOiIyMDI1MDMwMTA4NDQiLCJhbGciOiJSUzI1NiJ9.eyJpYW1faWQiOiJJQk1pZC02NjEwMDU0SUIxIiwiaWQiOiJJQk1pZC02NjEwMDU0SUIxIiwicmVhbG1pZCI6IklCTWlkIiwianRpIjoiZDY2YzU1ZmUtMzg3ZS00OWY3LWJhMjItM2MxMDNiYzEyMmMzIiwiaWRlbnRpZmllciI6IjY2MTAwNTRJQjEiLCJnaXZlbl9uYW1lIjoiQWFyb24gVGhvbWFzIiwiZmFtaWx5X25hbWUiOiJCbGVzc2VuIiwibmFtZSI6IkFhcm9uIFRob21hcyBCbGVzc2VuIiwiZW1haWwiOiJhYXJvbnQuY3NhMjEyNUBzYWludGdpdHMub3JnIiwic3ViIjoiYWFyb250LmNzYTIxMjVAc2FpbnRnaXRzLm9yZyIsImF1dGhuIjp7InN1YiI6ImFhcm9udC5jc2EyMTI1QHNhaW50Z2l0cy5vcmciLCJpYW1faWQiOiJJQk1pZC02NjEwMDU0SUIxIiwibmFtZSI6IkFhcm9uIFRob21hcyBCbGVzc2VuIiwiZ2l2ZW5fbmFtZSI6IkFhcm9uIFRob21hcyIsImZhbWlseV9uYW1lIjoiQmxlc3NlbiIsImVtYWlsIjoiYWFyb250LmNzYTIxMjVAc2FpbnRnaXRzLm9yZyJ9LCJhY2NvdW50Ijp7InZhbGlkIjp0cnVlLCJic3MiOiIyY2ZkZjBkNGUwYjk0NzRmYmJjNGQ0OTkxYWFmODkyOCIsImltc191c2VyX2lkIjoiMTM0MzI0MzgiLCJmcm96ZW4iOnRydWUsImltcyI6IjI5OTY4MzQifSwiaWF0IjoxNzQzMzU1NzMyLCJleHAiOjE3NDMzNTkzMzIsImlzcyI6Imh0dHBzOi8vaWFtLmNsb3VkLmlibS5jb20vaWRlbnRpdHkiLCJncmFudF90eXBlIjoidXJuOmlibTpwYXJhbXM6b2F1dGg6Z3JhbnQtdHlwZTphcGlrZXkiLCJzY29wZSI6ImlibSBvcGVuaWQiLCJjbGllbnRfaWQiOiJkZWZhdWx0IiwiYWNyIjoxLCJhbXIiOlsicHdkIl19.GshC_HY84bYC06RDwZF0Zgg6R_qhi1ZAXczPmFnr6YVke162aO2WJLteULYbyFbUfNLZduaz-1KaVY1RTEbISgleZbkmQzo9xa0585uErktuSIt4BBF56CKc5yNZ4D8lmC6n5qtY9yDDXWBq14L_gVdpp3mZWfDoLZSZh2pqbYiUGWJOZ3wLMdebbHSMOuwIKtCb-01AI5H56EHn06Cz6LDxv_tugn6IXm9jMg8Wx-6m2ECk09bOuE6XJQFP5_WyLwXwdGqUiWvmRoNRJrmp9GtSOZLfpOOURQwU0GocK6CzRkOfWq_HxVHOzrMpOUmTRVaQd-w1gPRW4mTRtWvctQ"
	};
	const body = {
		input: "You are an AI email assistant designed to intelligently manage email overload. Your responsibilities include:\n\n1️⃣ **Prioritization** - Determine urgency based on sender, subject, and content.  \n2️⃣ **Summarization** - Condense long email threads into key points.  \n3️⃣ **Response Suggestions** - Generate professional replies based on context.  \n4️⃣ **Automation** - Assist in scheduling, approvals, and follow-ups.\n\n### Instructions:\nAnalyze the provided email(s) and respond appropriately by selecting the most relevant task(s):  \n- **If the email is urgent, determine the Priority Level (High, Medium, Low).**  \n- **If it is a long email thread, summarize it into key points.**  \n- **If a response is needed, draft a professional reply.**  \n- **If the email requires action (e.g., scheduling, approvals), outline the next steps.**\n\n\nInput: **📝 Example 1: Email Prioritization**\n**Input Email:**  \nFrom: CEO ceo@company.com\nSubject: Urgent - Board Meeting Tomorrow\nBody:\nHi Team,\n\nWe need to finalize the agenda for tomorrow’s board meeting. Please send me the latest updates by 5 PM today.\n\nBest,\nCEO\nOutput: Priority Level: High\nReason: Email is from the CEO and requires urgent action before 5 PM today.\n\nInput: ### **📝 Example 2: Summarizing an Email Thread**\n**Input Email Thread:**  \n[Alice]: \"Hey team, let’s finalize the budget by next week.\"\n\n[Bob]: \"Agreed. I will send the financial breakdown tomorrow.\"\n\n[Alice]: \"Thanks, Bob. Once ready, let’s schedule a meeting with finance.\"\n\n[Carol]: \"I will coordinate and find a slot for us by Wednesday.\"\n\n[Bob]: \"Attached the budget doc for review.\"\n\n[Alice]: \"Looks good. Let’s proceed with the final approval process.\"\nOutput: Summary:\n\nThe team is finalizing the budget.\n\nBob provided a financial breakdown.\n\nCarol is scheduling a meeting by Wednesday.\n\nFinal approval process is in progress.\n\nyaml\nCopy\nEdit\n\n\nInput: ### **📝 Example 3: Suggesting a Response**\n**Input Email:**  \nFrom: Client client@company.com\nSubject: Proposal Review Request\nBody:\nHi [Your Name],\n\nWe have reviewed the proposal you sent last week. We would like some clarifications on the pricing model and implementation timeline. Can we set up a call this week?\n\nLooking forward to your response.\n\nBest,\nClient\nOutput: Suggested Response:\nSubject: Re: Proposal Review Request\n\nHi [Client Name],\n\nThank you for reviewing the proposal. I’d be happy to clarify the pricing model and implementation timeline. Could you share your availability for a call this week? I am available on [Provide available time slots].\n\nLooking forward to our discussion.\n\nBest regards,\n[Your Name]\n\nInput: ### **📝 Example 4: Automating Scheduling & Approvals**\n**Input Email:**  \nFrom: HR hr@company.com\nSubject: Leave Request Approval\nBody:\nHi [Manager Name],\n\nEmployee John Doe has submitted a leave request from March 25th to March 30th. Please review and approve if there are no conflicts.\n\nBest,\nHR\nOutput: Actionable Steps:\n\nCheck for conflicts in John Doe’s schedule.\n\nIf no conflicts, send approval response to HR.\n\nIf conflicts exist, propose alternate dates or escalate.\n\nSuggested Response:\nSubject: Re: Leave Request Approval\n\nHi HR,\n\nI have reviewed John Doe’s leave request for March 25th–30th. There are no scheduling conflicts, so I approve the request. Please proceed with the necessary updates.\n\nBest,\n[Manager Name]\n\nInput: From: VP Sales <vp_sales@company.com>  \nSubject: Urgent: Client Demo at 3 PM Today  \nBody:  \nHi Team,  \n\nThe client has requested a last-minute demo at 3 PM today. Please ensure everything is set up.  \n\nThanks,  \nVP Sales\n\nOutput:",
		parameters: {
			decoding_method: "greedy",
			max_new_tokens: 200,
			min_new_tokens: 0,
			stop_sequences: [],
			repetition_penalty: 1
		},
		model_id: "ibm/granite-13b-instruct-v2",
		project_id: "d4c45ea6-cd2e-4c05-b645-ffb438cf058e"
	};

	const response = await fetch(url, {
		headers,
		method: "POST",
		body: JSON.stringify(body)
	});

	if (!response.ok) {
		throw new Error("Non-200 response");
	}

	return await response.json();
}

generateText()
    .then(response => console.log(response))
    .catch(error => console.error("Error:", error));

File: Stream/Intelligent.py
import streamlit as st
import requests
import json
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

def analyze_email(user_email):
    url = "https://us-south.ml.cloud.ibm.com/ml/v1/text/generation?version=2023-05-29"
    headers = {
        "Accept": "application/json",
        "Content-Type": "application/json",
        "Authorization": f"Bearer {os.getenv('IBM_API_TOKEN')}"
    }
    body = {
        "input": f"""You are an AI email assistant designed to intelligently manage email overload. Your responsibilities include:

1️⃣ **Prioritization** - Determine urgency based on sender, subject, and content.  
2️⃣ **Summarization** - Condense long email threads into key points.  
3️⃣ **Response Suggestions** - Generate professional replies based on context.  
4️⃣ **Automation** - Assist in scheduling, approvals, and follow-ups.

Input: {user_email}
Output:""",
        "parameters": {
            "decoding_method": "greedy",
            "max_new_tokens": 200,
            "min_new_tokens": 0,
            "stop_sequences": [],
            "repetition_penalty": 1
        },
        "model_id": "ibm/granite-13b-instruct-v2",
        "project_id": "d4c45ea6-cd2e-4c05-b645-ffb438cf058e"
    }

    try:
        response = requests.post(url, headers=headers, json=body)
        response.raise_for_status()  # Raise exception for non-200 responses
        return response.json()
    except Exception as e:
        st.error(f"Error: {str(e)}")
        return None

def main():
    st.set_page_config(
        page_title="Email Assistant",
        page_icon="📧",
        layout="wide"
    )

    st.title("AI Email Assistant")
    st.markdown("""
    This application helps you manage email overload by:
    * **Prioritizing** emails based on urgency
    * **Summarizing** long email threads
    * **Suggesting** professional responses
    * **Automating** scheduling and follow-ups
    """)

    # Example emails for quick selection
    st.sidebar.title("Example Emails")
    
    example_emails = {
        "Urgent Meeting Request": """From: Manager manager@company.com
Subject: Emergency Team Meeting
Body: We need to discuss an urgent client issue. Please join a call in 30 minutes.""",
        
        "Project Update Thread": """[Tom]: Project status update - we're on track for delivery
[Sarah]: Great progress! Few concerns about the testing phase
[Tom]: We can extend testing by 2 days if needed
[Sarah]: Agreed, let's update the timeline""",
        
        "Client Feedback Email": """From: Client client@company.com
Subject: Product Feedback
Body: The recent updates are good but we noticed some performance issues.
Could we schedule a call to discuss?"""
    }
    
    for example_name, example_text in example_emails.items():
        if st.sidebar.button(example_name):
            st.session_state.email_text = example_text
    
    # Custom email input
    if 'email_text' not in st.session_state:
        st.session_state.email_text = ""
    
    email_text = st.text_area(
        "Enter your email to analyze:", 
        value=st.session_state.email_text,
        height=200
    )
    
    col1, col2 = st.columns([1, 5])
    with col1:
        analyze_button = st.button("Analyze Email", type="primary")
    with col2:
        clear_button = st.button("Clear")
        
    if clear_button:
        st.session_state.email_text = ""
        st.experimental_rerun()
    
    # Process email and display results
    if analyze_button and email_text:
        with st.spinner('Analyzing your email...'):
            response = analyze_email(email_text)
            
        if response and 'results' in response:
            st.success("Analysis Complete")
            
            # Extract and display the analysis
            analysis_text = response['results'][0]['generated_text'].strip()
            
            # Create tabs for different types of analysis
            tab1, tab2, tab3 = st.tabs(["Complete Analysis", "Summary", "Response Suggestion"])
            
            with tab1:
                st.markdown("### Complete Analysis")
                st.markdown(analysis_text)
            
            with tab2:
                st.markdown("### Summary")
                # Try to extract summary from the analysis
                if "Summary:" in analysis_text:
                    summary_start = analysis_text.find("Summary:")
                    summary_end = analysis_text.find("\n\n", summary_start)
                    if summary_end == -1:  # If no double newline found
                        summary_end = len(analysis_text)
                    summary = analysis_text[summary_start:summary_end].strip()
                    st.markdown(summary)
                else:
                    st.markdown("Detailed summary not available for this email.")
            
            with tab3:
                st.markdown("### Suggested Response")
                # Try to extract response suggestion from the analysis
                if "Suggested Response:" in analysis_text:
                    response_start = analysis_text.find("Suggested Response:")
                    response_text = analysis_text[response_start:].strip()
                    st.markdown(response_text)
                    
                    # Add a button to copy response to clipboard (note: this uses JavaScript)
                    response_only = response_text.replace("Suggested Response:", "").strip()
                    st.code(response_only)
                    st.button("Copy to Clipboard", 
                              on_click=lambda: st.write("Response copied to clipboard!"))
                else:
                    st.markdown("Response suggestion not available for this email.")
            
            # Optional: Show raw JSON response
            with st.expander("View Raw API Response"):
                st.json(response)
        else:
            st.error("Failed to analyze the email. Please try again.")

if __name__ == "__main__":
    main()